lib/matplotlib/backends/backend_cairo.py

```diff
@@ -11,8 +11,7 @@
 
 import numpy as np
 
-# cairocffi is more widely compatible than pycairo (in particular pgi only
-# works with cairocffi) so try it first.
+# cairocffi is more widely compatible than pycairo so try it first.
```

Comment:
Perhaps it makes sense to prefer pycairo now?  (as pygobject will go through pycairo regardless).
In fact, does pygobject even work (in this use case) without pycairo?

Reply:
I saw that there is code to convert pycairo object to cairocffi ones for some fastpath implementations, so I assumed it is still preferred.

If there is anything I can do in pycairo to make it the preferred one I'm happy to help.

Reply:
Which means in particular on_draw_event (connected to the "draw" event), no?
IOW, can we just assume that gtk3 is built with cairo support and that cairo is a dependency, and error out early if not?

I don't think keeping the cairocffi optional dependency just for the sake of a not-even-twofold speedup is worth it (but if you want to make expose something similar in pycairo's side what would be needed is a way to create a cairo Path from a list of cairo_path_data_t (or better, from a buffer (in the memoryview sense), so that no copy at all is done).  (Effectively, this runs afoul of the last paragraph of https://cairographics.org/manual/bindings-path.html ("You should not present an API for mutating or for creating new cairo_path_t objects. In the future, these guidelines may be extended to present an API for creating a cairo_path_t from scratch for use with cairo_append_path() but the current expectation is that cairo_append_path() will mostly be used with paths from cairo_copy_path()."), but performance wise it does make some difference.)

Reply:
> Which means in particular on_draw_event (connected to the "draw" event), no?

yeah

>  IOW, can we just assume that gtk3 is built with cairo support and that cairo is a dependency, and error out early if not?

We have https://pygobject.readthedocs.io/en/latest/guide/api/api.html?highlight=require_foreign#gi.require_foreign to check for this.

> I don't think keeping the cairocffi optional dependency just for the sake of a not-even-twofold speedup is worth it [...]

Thanks, I'll have a look.

----

Either way I think any changes to the cairo part should be another PR.

Reply:
It was originally done to support py3 but I guess yeah I don't know the situation with pypy right now either, @lazka?

Reply:
xenial has pycairo 1.10 but we require 1.11 indeed.

Reply:
@anntzer if I just take the code in _append_paths_fast() that computes codes/vertices and loop over them and call the path functions I get the same FPS boost as with the buffer passing approach (unless I'm missing something). So I think the while cairo_path_data_t thing is not needed and the fast path would also works with pycairo as is.

Reply:
ah, the "fast path" cleans up the path first, so they aren't doing the same thing. The main speedup comes from removing the curve segments. So assuming that shortcut is OK this would be faster:
```python
# Mapping from Matplotlib Path codes to cairo path functions.
_MPL_TO_CAIRO_PATH_TYPE = [lambda ctx, *args: None] * (Path.CLOSEPOLY + 1)
_MPL_TO_CAIRO_PATH_TYPE[Path.MOVETO] = cairo.Context.move_to
_MPL_TO_CAIRO_PATH_TYPE[Path.LINETO] = cairo.Context.line_to
_MPL_TO_CAIRO_PATH_TYPE[Path.CLOSEPOLY] = \
    lambda ctx, *args: cairo.Context.close_path(ctx)


def _append_paths(ctx, paths, transforms, clip=None):
    for path, transform in zip(paths, transforms):
        cleaned = path.cleaned(transform, remove_nans=True, clip=clip)
        for points, code in zip(cleaned.vertices, cleaned.codes):
            _MPL_TO_CAIRO_PATH_TYPE[code](ctx, *points)
```